File: solutions/clojure/045-insert-interval.clj
;; Problem 45: Insert Interval (LeetCode 57)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn insert-interval [intervals new-interval]
  (let [all (sort-by first (conj intervals new-interval))]
    (reduce (fn [acc [s e]]
              (if-let [[ps pe] (peek acc)]
                (if (<= s pe)
                  (conj (pop acc) [ps (max pe e)])
                  (conj acc [s e]))
                (conj acc [s e])))
            [] all)))

File: solutions/ocaml/045-insert-interval.ml
(*
  Problem 45: Insert Interval (LeetCode 57)
  Difficulty: Med
  Language: OCaml
*)
let insert_interval intervals (ns, ne) =
  let all = List.sort (fun a b -> compare (fst a) (fst b)) ((ns, ne) :: intervals) in
  List.fold_left (fun acc (s, e) ->
    match acc with
    | (ps, pe) :: rest when s <= pe -> (ps, max pe e) :: rest
    | _ -> (s, e) :: acc
  ) [] all |> List.rev
